src/V3HierBlock.cpp

```diff
@@ -0,0 +1,390 @@
+//    (which block uses other blocks..)
+// 8) In V3EmitMk.cpp, ${prefix}_hier.mk is created in run a).
+//
+// There are two hidden command options.
+//   --hierarchical-mode is added to Verilator run b).
+//   --hierarchy-block module_name,mangled_name,name0,value0,name1,value1,...
+//       module_name  :The original modulename
+//       mangled_name :Mangled name of parameterized modules (named in V3Param.cpp).
+//                     Same as module_name for non-parameterized hierarchy block.
+//       name         :The name of the parameter
+//       value        :Overridden value of the parameter
+//
+//       Used for b) and c).
+//       This options is repeated for all instantiating hierarchy blocks.
+
+#include <memory>
+#include <sstream>
+#include <utility>
+#include <vector>
+
+#include "V3Ast.h"
+#include "V3Error.h"
+#include "V3HierBlock.h"
+#include "V3Stats.h"
+
+typedef std::vector<std::pair<string, string> > StrGParams;
+
+static string escapeIntegerForShell(const string& str) {
+    string result;
+    result.reserve(str.size() + 1);
+    for (size_t i = 0; i < str.size(); ++i) {
+        if (str[i] == '\'') result.push_back('\\');
+        result.push_back(str[i]);
+    }
+    return result;
+}
+
+static string escapeStringForShell(const string& str) {
+    string result;
+    const char squoate = '\'';
+    const char dquoate = '"';
+    result.reserve(str.size() + 2);
+    // String literal for -G option must be surrounded by double quote (")
+    result.push_back(squoate);
+    result.push_back(dquoate);
+    result.push_back(squoate);
+
+    result.push_back(squoate);
+    for (size_t i = 0; i < str.size(); ++i) {
+        if (str[i] == squoate) {
+            result.push_back(squoate);  // terminate
+            result.push_back(dquoate);  // "
+            result.push_back(squoate);  // str[i]
+            result.push_back(dquoate);  // "
+            result.push_back(squoate);  // start
+        } else {
+            result.push_back(str[i]);
+        }
+    }
+    result.push_back(squoate);
+
+    result.push_back(squoate);
+    result.push_back(dquoate);
+    result.push_back(squoate);
+    return result;
+}
+
+static string escapeEscape(const string& s) {
+    string result;
+    result.reserve(s.length());
+    for (string::const_iterator it = s.begin(); it != s.end(); ++it) {
+        if (*it == '\\') result.push_back(*it);
+        result.push_back(*it);
+    }
+    return result;
+}
+
+static string escapeStringForCMake(const string& str) {
+    int num_eq = 0;
+    for (string::size_type i = 0; i < str.length(); ++i) {
+        if (str[i] == '=') ++num_eq;
+    }
+    string eq;
+    eq.resize(num_eq, '=');
+    return '[' + eq + '[' + str + ']' + eq + ']';
+}
+
+static StrGParams stringifyParams(const V3HierBlock::GParams& gparams, bool forCMake,
+                                  bool forGOption) {
+    StrGParams strParams;
```

Comment:
How does this process handle parameters that aren't simple constants - e.g. "parameter type"?  Please add a parameter type to your tests.